velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -51,6 +50,48 @@ class PrefixSortEncoder {
     }
   }
 
+  /// Encodes String types.
+  /// The first byte of the encoded result is null byte. The value is 0 if
+  /// (nulls first and value is null) or (nulls last and value is not null).
+  /// Otherwise, the value is 1.
+  FOLLY_ALWAYS_INLINE void encode(
+      std::optional<StringView> value,
+      char* dest,
+      int32_t encodeSize) const {
+    if (value.has_value()) {
+      dest[0] = nullsFirst_ ? 1 : 0;
+      encodeNoNulls(value.value(), dest + 1, encodeSize - 1);
+    } else {
+      dest[0] = nullsFirst_ ? 0 : 1;
+      std::memset(dest + 1, 0, encodeSize - 1);
+    }
+  }
+
+  /// The string prefix is formatted as 'null byte + string content + padding
+  /// zeros'. If `!ascending_`, the bits for both the content and padding zeros
+  /// need to be inverted.
+  FOLLY_ALWAYS_INLINE void encodeNoNulls(
+      std::optional<StringView> value,
+      char* dest,
+      int32_t encodeSize) const {
+    std::string storage;
+    // 'value' may be stored in non-contiguous allocation pieces in the row
+    // container, so we need to use 'contiguousString' method to read it out.
+    auto data = HashStringAllocator::contiguousString(value.value(), storage);
```

Comment:
auto stream = HashStringAllocator::prepareRead(
      HashStringAllocator::headerOf(value.data()));
  stream->readBytes(dest, value.size());
to copy the value once.

It refers to RowContainer::extractString

Reply:
Then we can only copy the prefix, not the total data.

Reply:
Make sense, updated.